internal/gitprovider/registry.go

```diff
@@ -0,0 +1,50 @@
+package gitprovider
+
+import (
+	"fmt"
+)
+
+// ProviderRegistration holds details on how to instantiate the correct git provider
+// based on parameters (i.e. repo URL). It allows GitProviderService implementations
+// to generically register themselves to
+type ProviderRegistration struct {
+	// Predicate is a function which should return true if the given repoURL is appropriate
+	// for the provider to handle (e.g. github.com is the domain name)
+	Predicate func(repoURL string) bool
+	// NewService instantiates the git provider
+	NewService func() (GitProviderService, error)
+}
+
+var (
+	// registeredProviders is a mapping between provider name and provider registration
+	registeredProviders = map[string]ProviderRegistration{}
+)
+
+// NewGitProviderServiceFromName returns a git provider service by it's registered name
+func NewGitProviderServiceFromName(name string) (GitProviderService, error) {
+	if reg, ok := registeredProviders[name]; ok {
+		return reg.NewService()
+	}
+	return nil, fmt.Errorf("No registered providers with name %q", name)
+}
```

Comment:
This appears to be unused. Is there a scenario where we need to find the correct provider implementation by name rather than by URL?

Reply:
Oh yes, I added this even though it is unused because once we have 2+ providers, none of them using publically hosted well-known URLs (github.com, gitlab.com, bitbucket.com, etc...), then users can explicitly instantiate it by name:

This syntax is actually already added in the spec:

```yaml
    gitRepoUpdates:
    - pullRequest:
        github: {}
```

However, it's not yet used because we don't have two providers and so there wasn't a need to instantiate it by name. 

That said, I think I should add the code to instantiate by name, so that the next person to add support for the next provider will know what to do. I will make that change.